contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,274 @@
+order to have vendors advertise and monitor their resources on behalf
+of Kubelet.
+
+Additionally, we introduce the concept of ResourceType to be able to select
+resources with constraints in a pod spec.
+
+_GPU Integration Example:_
+  * [Enable "kick the tires" support for Nvidia GPUs in COS](https://github.com/kubernetes/kubernetes/pull/45136)
+  * [Extend experimental support to multiple Nvidia GPUs](https://github.com/kubernetes/kubernetes/pull/42116)
+
+_Kubernetes Meeting Notes On This:_
+  * [Meeting notes](https://docs.google.com/document/d/1Qg42Nmv-QwL4RxicsU2qtZgFKOzANf8fGayw8p3lX6U/edit#)
+  * [Better Abstraction for Compute Resources in Kubernetes](https://docs.google.com/document/d/1666PPUs4Lz56TqKygcy6mXkNazde-vwA7q4e5H92sUc)
+  * [Extensible support for hardware devices in Kubernetes (join kubernetes-dev@googlegroups.com for access)](https://docs.google.com/document/d/1LHeTPx_fWA1PdZkHuALPzYxR0AYXUiiXdo3S0g2VSlo/edit)
+
+## Use Cases
+
+  * I want to use a particular device type (GPU, InfiniBand, FPGA, etc.)
+    in my pod.
+  * I should be able to use that device without writing custom Kubernetes code.
+  * I want a consistent and portable solution to consuming hardware devices
+    across k8s clusters
+
+## Objectives
+
+1. Create a plugin mechanism which allows discovery and monitoring of devices
+2. Add support for ResourceType in the scheduler and kubelet
+3. Define a deployment mechanism
+4. Versioning the new interface/API
+
+## Non Objectives
+1. Driver Installation which should be addressed in a follow up PR. In theory
+   the Device Plugin can work if we assume that the drivers are already installed
+   in the machine.
+2. Metrics this should be the job of cadvisor and should probably either addressed
+   there or if people feel there is a case to be made for it being addressed in the
+   Device Plugin in a follow up proposal.
+
+## Proposed changes
+
+### Installation
+
+The installation process should be straightforward to the user, transparent
+and similar to other regular Kubernetes actions.
+The device plugin should also run in containers so that kubernetes can
+deploy them and restart the plugins when they fail.
+
+Following this, it makes sense that the user would deploy Device Plugins
+through DaemonSets.
+
+For users installing Kubernetes without using an integrated solution
+such as `kubeadm` they could use the examples that we would put in
+`https://github.com/kubernetes/kubernetes/tree/master/examples/device-plugin`
+
+In the case of integrated solutions such as `kubeadm` we can have them support
+auto-deploying of vetted Device Plugins.
+
+### API Changes
+#### ResourceType
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called the ResourceType.
+It is defined as follows:
+
+```golang
+type ResourceType struct {
+	Kind string
+	Name string
```

Comment:
exactly!!
I am working on [Resource Class proposal.](https://docs.google.com/document/d/1-BAR-RSFkXLbqRREW6jE7JGipmGi2ygWxbwIZ8Eq_NQ/edit#heading=h.gnxpzs578dw7) For now i am hacking the device discovery part and updating fake device in node status. Since both the proposals, device plugin and resource class, share a common structure Device, which is updated by device plugin and consumed by resource class proposal, we should agree on same data structure.

**Example on subResources**
Suppose there is a device plugin for nvidia-gpu devices and this device plugin wants to advertise two category of devices:
* Just a single nvidia GPU
* A group of nvidia GPUs grouped for better performance. Ex, shared PCI bus or nvlink connectivity between them etc.
To represent a device like described in 2,  `Device` object need to have `subResources`, where subResources will just have `name` and  `quantity`. `quantity` is the number of devices in grouping.
`type` for both the devices,  grouped and subResource, will be the same, ex `nvidia-gpu`.
Their names will differ, for example `nvidia-gpu` and `nvidia-gpu-nvlinked-group-of-2`
Property by which subResource devices are grouped to create another device like the one described in 2,  needs to be set to `true` for both the devices, subResource device and grouped device.
 
If a Resource Class wants to represent a device like described in 2 above, it must have:
 * Two key-value pairs in device properties list, (k,v,op), `type=<device-type>` and `<grouping-deciding-charecteristic>=true` ex:
               type=nvidia-gpu
               nvlink=true
* subResourceCount
subResourceCount should be a specific member in `Device` structure because it will standardize the device parsing logic across different device plugins. 
Make sense?